### PR TITLE
Use suitesparse QR for sparse GP models [OC-310]

### PR DESCRIPTION
**Pull request by peddie**
When the observation matrix (covariance between measurements to inducing points) is sufficiently sparse, a fit or update to a `SparseGP` model can be computed significantly more quickly using a sparse solver.  This patch adds support for doing this calculation with the multifrontal SPQR solver from Suitesparse rather than with Eigen's built-in Householder QR routine.

## Measurements

On a Swift model with fill ratio around 0.3, around 10k measurements and 1k inducing points, adopting SPQR alone provided over 4x speedup in the _O(N^3)_ solve step.  This increased to nearly 7x when SPQR was instructed to use 8 threads internally.  In combination with MKL, the speedup vs. the original dense QR approached 9x (again with 8 threads).

Your results may vary depending on your problem, architecture and more.

### CODE REVIEW

File: include/albatross/src/models/sparse_gp.hpp [hunk 1]
```diff
@@ -370,9 +370,10 @@ class SparseGaussianProcessRegression
   // };
 
   // TODO(@peddie) SPQR configuration
```
Comment (akleeman) [inline include/albatross/src/models/sparse_gp.hpp:372]: what does this TODO mean?

Comment (peddie) [reply include/albatross/src/models/sparse_gp.hpp:372]: Hmm . . . that's a good question.  I think it's outdated, since `SPQR_create()` sets pivot tolerance and thread count.

File: include/albatross/src/cereal/eigen.hpp [hunk 2]
```diff
@@ -147,6 +147,74 @@ inline void load_lower_triangle(Archive &archive,
...
+  for (_StorageIndex col = 0; col < m.outerSize(); ++col) {
+    for (typename Eigen::SparseMatrix<_Scalar, _Options,
+                                      _StorageIndex>::InnerIterator it(m, col);
+         it; ++it) {
+      elems.emplace_back(it.col(), it.row(), it.value());
+    }
+  }
+
+  std::string payload =
+      gzip::compress(reinterpret_cast<const char *>(elems.data()), size_bytes);
```
Comment (akleeman) [inline include/albatross/src/cereal/eigen.hpp:170]: I guess we can followup, but curious if this should end up using zstd?

File: include/albatross/src/cereal/eigen.hpp [hunk 3]
```diff
@@ -147,6 +147,74 @@ inline void load_lower_triangle(Archive &archive,
...
+  ar(CEREAL_NVP(nnz));
+  ar(CEREAL_NVP(payload));
+  const std::size_t size_bytes = nnz * sizeof(_Scalar);
+
+  if (::cereal::traits::is_text_archive<Archive>::value) {
+    payload = base64::decode(payload);
+  }
+
+  const std::string decompressed =
+      gzip::decompress(payload.data(), payload.size());
```
Comment (john-michaelburke) [inline include/albatross/src/cereal/eigen.hpp:202]: At the risk of adding more work for you, when you have clearly put a lot of work into this PR already :grimacing: 

It seems like in another PR you have up, you switched to zstd compression in place of gzip. Are the benefits of gzip over zstd situational or was there another reason? I saw in your other PR you used a zstd::maybe_decompress and then would fall back on gzip, is that something we could do here and force all compression to use zstd?

**Review by john-michaelburke — APPROVED** [review]

File: include/albatross/src/cereal/eigen.hpp (hunk 3 above)
Comment (peddie) [reply include/albatross/src/cereal/eigen.hpp:202]: This just uses gzip because it's not rebased on the other change.  Depending on which goes in first, I'll update this code accordingly.  We will use zstd in the end.

File: include/albatross/src/cereal/eigen.hpp [hunk 4]
```diff
@@ -147,6 +147,74 @@ inline void load_lower_triangle(Archive &archive,
...
+  elems.reserve(nnz);
+  for (_StorageIndex col = 0; col < m.outerSize(); ++col) {
+    for (typename Eigen::SparseMatrix<_Scalar, _Options,
+                                      _StorageIndex>::InnerIterator it(m, col);
+         it; ++it) {
+      elems.emplace_back(it.col(), it.row(), it.value());
+    }
+  }
+
+  std::string payload =
```
Comment (akleeman) [inline include/albatross/src/cereal/eigen.hpp:169]: encode_array work here?

File: include/albatross/src/cereal/suitesparse.hpp [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+  for (std::size_t m = 0; m < max_methods; ++m) {
+    ar(cc.method[m]);
+  }
+  // method
+  ar(CEREAL_NVP(cc.postorder));
+  ar(CEREAL_NVP(cc.default_nesdis));
+  ar(CEREAL_NVP(cc.metis_memory));
+  ar(CEREAL_NVP(cc.metis_dswitch));
+  ar(CEREAL_NVP(cc.metis_nswitch));
+  ar(::cereal::make_nvp("nrow", cc.nrow));
```
Comment (akleeman) [inline include/albatross/src/cereal/suitesparse.hpp:147]: why does it sometimes use `CEREAL_NVP` and othertimes explicit names?

File: include/albatross/src/cereal/suitesparse.hpp [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+  const std::size_t cholmod_int_size =
+      cc.itype == CHOLMOD_LONG ? sizeof(SuiteSparse_long) : sizeof(int);
+
+  detail::encode_array(ar, "cc.Flag", cc.Flag, cc.nrow * cholmod_int_size);
+  detail::encode_array(ar, "cc.Head", cc.Head,
+                       (cc.nrow + 1) * cholmod_int_size);
+  detail::encode_array(ar, "cc.Xwork", cc.Xwork, cc.xworksize * sizeof(double));
+
+  // Iwork omitted; must reinitialize
```
Comment (akleeman) [inline include/albatross/src/cereal/suitesparse.hpp:162]: Not sure I follow this comment?

File: include/albatross/src/cereal/suitesparse.hpp [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+
+  detail::encode_array(ar, "m.p", m.p, (m.ncol + 1) * p_size_bytes);
+  detail::encode_array(ar, "m.i", m.i, m.nzmax * integer_size_bytes);
+  if (nullptr != m.nz && !m.packed) {
+    detail::encode_array(ar, "m.nz", m.nz, m.ncol * integer_size_bytes);
+  }
+  if (nullptr != m.x) {
+    detail::encode_array(ar, "m.x", m.x, x_elements * element_size_bytes);
+  }
+  // This should only be non-null for "zomplex" elements
```
Comment (akleeman) [inline include/albatross/src/cereal/suitesparse.hpp:401]: complex?

File: include/albatross/src/cereal/suitesparse.hpp [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+    detail::suitesparse_cereal_realloc(&m.nz, m.ncol, integer_size_bytes);
+  }
+  detail::suitesparse_cereal_realloc(&m.x, m.nzmax, element_size_bytes);
+
+  detail::decode_array(ar, "m.p", m.p, (m.ncol + 1) * p_size_bytes);
+  detail::decode_array(ar, "m.i", m.i, m.nzmax * integer_size_bytes);
+  if (!m.packed) {
+    detail::decode_array(ar, "m.nz", m.nz, m.ncol * integer_size_bytes);
+  }
+  // TODO(@peddie): in what cases may `x` be absent?
```
Comment (akleeman) [inline include/albatross/src/cereal/suitesparse.hpp:435]: is this a concern? should we add a ticket to follow up?

File: include/albatross/src/models/sparse_gp.hpp [hunk 9]
```diff
@@ -317,7 +319,8 @@ class SparseGaussianProcessRegression
       B.col(pi).topRows(i + 1) = old_fit.sigma_R.col(i).topRows(i + 1);
     }
     B.bottomRows(n_new) = A_ldlt.sqrt_solve(K_fu);
-    const auto B_qr = B.colPivHouseholderQr();
+    auto B_qr = SPQR_create(B.sparseView());
```
Comment (akleeman) [inline include/albatross/src/models/sparse_gp.hpp]: curious how long `SPQR_create` takes, would it be helpful if `B` were already a sparse matrix type?

File: include/albatross/src/models/sparse_gp.hpp [hunk 10]
```diff
@@ -350,14 +359,25 @@ class SparseGaussianProcessRegression
   //
   //   Sigma = (B^T B)^-1
   //
-  Eigen::ColPivHouseholderQR<Eigen::MatrixXd>
+  // Eigen::ColPivHouseholderQR<Eigen::MatrixXd>
```
Comment (akleeman) [inline include/albatross/src/models/sparse_gp.hpp:362]: delete?

**Review by akleeman — COMMENTED** [review]
mostly looks good, handful of nits/questions.

I guess we just need to trust unit tests to know everything was serialized. Are there any other edge cases on the matrix usage that we should check?

Very curious to see how this changes run time on some of our use cases.

**Comment by akleeman** [comment]
could use an improved description

File: include/albatross/src/cereal/eigen.hpp (hunk 2 above)
Comment (peddie) [reply include/albatross/src/cereal/eigen.hpp:170]: See JM's comment above!

File: include/albatross/src/cereal/suitesparse.hpp (hunk 7 above)
Comment (peddie) [reply include/albatross/src/cereal/suitesparse.hpp:401]: no . . . sadly "zomplex" is an internal suitesparse representation.

File: include/albatross/src/cereal/suitesparse.hpp (hunk 8 above)
Comment (peddie) [reply include/albatross/src/cereal/suitesparse.hpp:435]: I will check.

File: include/albatross/src/cereal/eigen.hpp (hunk 4 above)
Comment (peddie) [reply include/albatross/src/cereal/eigen.hpp:169]: It should -- I could expose that to this code; so far it's local to the suitesparse serialisation code.

File: include/albatross/src/cereal/suitesparse.hpp (hunk 5 above)
Comment (peddie) [reply include/albatross/src/cereal/suitesparse.hpp:147]: In most of the occurrences, this happens because I'm serialising elements of a fixed-length array.

The reason it happens this way for `nrow`, `iworksize` and `xworksize` is that I want to deserialise them to stack variables so I can calculate allocation sizes etc., rather than deserialising them to elements of `cholmod_common` directly (because initialisation functions can and do modify various elements of `cholmod_common` structures).  I'll add a comment here.

File: include/albatross/src/cereal/suitesparse.hpp (hunk 6 above)
Comment (peddie) [reply include/albatross/src/cereal/suitesparse.hpp:162]: This is related to the previous thing -- I don't serialise the workspaces, I just reallocate new ones when I deserialise.  (I confirmed that this is safe both with tests and by reading the SPQR / cholmod code.)  I'll expand the comment to make it clearer.

File: include/albatross/src/models/sparse_gp.hpp (hunk 9 above)
Comment (peddie) [reply include/albatross/src/models/sparse_gp.hpp]: It would definitely be helpful, but it significantly complicates assembling `B`, as you can't just assign to block views of sparse matrices.  I don't mind making that change if you'd like to see it.

I haven't done any benchmarks on this whole integrated system yet; my benchmarks were just on solving sparse problems I'd dumped out from a model already in sparse format.  I'll put some head-to-head test results up for the Swift model that uses a sparse GP in that repo.

File: include/albatross/src/models/sparse_gp.hpp (hunk 10 above)
Comment (peddie) [reply include/albatross/src/models/sparse_gp.hpp:362]: Yes, thanks!

**Comment by akleeman** [comment]
How crazy would it be to copy the `sparse_gp.hpp` to a `suitesparse_gp.hpp` (or something like that) to make it easy to switch?

**Comment by peddie** [comment]
> How crazy would it be to copy the `sparse_gp.hpp` to a `suitesparse_gp.hpp` (or something like that) to make it easy to switch and compare performance?

Check the most recent patch.  The changes are much smaller than that.

File: include/albatross/src/models/sparse_gp.hpp [hunk 11]
```diff
@@ -317,7 +341,9 @@ class SparseGaussianProcessRegression
       B.col(pi).topRows(i + 1) = old_fit.sigma_R.col(i).topRows(i + 1);
     }
     B.bottomRows(n_new) = A_ldlt.sqrt_solve(K_fu);
-    const auto B_qr = B.colPivHouseholderQr();
+    auto B_qr = QRImplementation::compute(
+        B, Base::threads_
+               .get()); // SPQR_create(B.sparseView(), Base::threads_.get());
```
Comment (akleeman) [inline include/albatross/src/models/sparse_gp.hpp:346]: nit: remove comment. Should `B_qr` be `const`?

File: include/albatross/src/models/sparse_gp.hpp [hunk 12]
```diff
@@ -349,15 +382,14 @@ class SparseGaussianProcessRegression
...
-  Eigen::ColPivHouseholderQR<Eigen::MatrixXd>
+  std::unique_ptr<Eigen::SPQR<Eigen::SparseMatrix<double>>>
   compute_sigma_qr(const Eigen::SerializableLDLT &K_uu_ldlt,
                    const BlockDiagonalLDLT &A_ldlt,
                    const Eigen::MatrixXd &K_fu) const {
     Eigen::MatrixXd B(A_ldlt.rows() + K_uu_ldlt.rows(), K_uu_ldlt.rows());
     B.topRows(A_ldlt.rows()) = A_ldlt.sqrt_solve(K_fu);
     B.bottomRows(K_uu_ldlt.rows()) = K_uu_ldlt.sqrt_transpose();
-    return B.colPivHouseholderQr();
+    return SPQR_create(B.sparseView(), Base::threads_.get());
```
Comment (akleeman) [inline include/albatross/src/models/sparse_gp.hpp:392]: So we're keeping the QR of sigma sparse for both implementations? or should this be using the `QRImplementation`?

Comment (peddie) [reply include/albatross/src/models/sparse_gp.hpp:392]: Oops, I updated the update but not the initial fit!

File: include/albatross/src/models/sparse_gp.hpp (hunk 11 above)
Comment (peddie) [reply include/albatross/src/models/sparse_gp.hpp]: Done.

**Comment by akleeman** [comment]
Mind adding a description of the changes (maybe some of the timing comparison plots from the first prototype?) for any future developers.
